File: backend/cmd/crawl.go
package cmd

import (
	"errors"
	"fjnkt98/atcodersearch/atcoder"
	"fjnkt98/atcodersearch/crawl"
	"fjnkt98/atcodersearch/repository"
	"fmt"
	"strings"
	"time"

	"github.com/urfave/cli/v2"
)

var ErrNoAuthInfo = errors.New("no auth info")

func NewCrawlCmd() *cli.Command {
	return &cli.Command{
		Name: "crawl",
		Flags: []cli.Flag{
			&cli.DurationFlag{
				Name:    "duration",
				Aliases: []string{"d"},
				Value:   1000 * time.Millisecond,
			},
		},
		Subcommands: []*cli.Command{
			{
				Name: "problem",
				Flags: []cli.Flag{
					&cli.BoolFlag{
						Name:    "all",
						Aliases: []string{"a"},
						Value:   false,
					},
				},
				Action: func(ctx *cli.Context) error {
					atcoderClient, err := atcoder.NewAtCoderClient()
					if err != nil {
						return fmt.Errorf("crawl problem: %w", err)
					}
					problemsClient := atcoder.NewAtCoderProblemsClient()

					pool, err := repository.NewPool(ctx.Context, ctx.String("database-url"))
					if err != nil {
						return fmt.Errorf("crawl problem: %w", err)
					}

					crawler := crawl.NewProblemCrawler(atcoderClient, problemsClient, pool, ctx.Duration("duration"), ctx.Bool("all"))
					if err := crawler.CrawlContests(ctx.Context); err != nil {
						return fmt.Errorf("crawl problem: %w", err)
					}
					if err := crawler.CrawlDifficulties(ctx.Context); err != nil {
						return fmt.Errorf("crawl problem: %w", err)
					}
					if err := crawler.CrawlProblems(ctx.Context); err != nil {
						return fmt.Errorf("crawl problem: %w", err)
					}

					return nil
				},
			},
			{
				Name: "user",
				Action: func(ctx *cli.Context) error {
					atcoderClient, err := atcoder.NewAtCoderClient()
					if err != nil {
						return fmt.Errorf("crawl user: %w", err)
					}

					pool, err := repository.NewPool(ctx.Context, ctx.String("database-url"))
					if err != nil {
						return fmt.Errorf("crawl user: %w", err)
					}

					crawler := crawl.NewUserCrawler(atcoderClient, pool, ctx.Duration("duration"))
					if err := crawler.Crawl(ctx.Context); err != nil {
						return fmt.Errorf("crawl user: %w", err)
					}

					return nil
				},
			},
			{
				Name: "submission",
				Flags: []cli.Flag{
					&cli.IntFlag{
						Name:    "retry",
						Aliases: []string{"r"},
						Value:   0,
					},
					&cli.StringFlag{
						Name: "target",
					},
					&cli.BoolFlag{
						Name: "endless",
					},
					&cli.StringFlag{
						Name:    "atcoder-username",
						Hidden:  true,
						EnvVars: []string{"ATCODER_USERNAME"},
					},
					&cli.StringFlag{
						Name:    "atcoder-password",
						Hidden:  true,
						EnvVars: []string{"ATCODER_PASSWORD"},
					},
				},
				Action: func(ctx *cli.Context) error {
					atcoderClient, err := atcoder.NewAtCoderClient()
					if err != nil {
						return fmt.Errorf("crawl submission: %w", err)
					}
					username := ctx.String("atcoder-username")
					password := ctx.String("atcoder-password")
					if username == "" || password == "" {
						return ErrNoAuthInfo
					}
					if err := atcoderClient.Login(ctx.Context, username, password); err != nil {
						return fmt.Errorf("login: %w", err)
					}
					pool, err := repository.NewPool(ctx.Context, ctx.String("database-url"))
					if err != nil {
						return fmt.Errorf("crawl submission: %w", err)
					}

					var targets []string
					if target := ctx.String("target"); target != "" {
						targets = strings.Split(target, ",")
					}

					crawler := crawl.NewSubmissionCrawler(
						atcoderClient,
						pool,
						ctx.Duration("duration"),
						ctx.Int("retry"),
						1*time.Minute,
						targets,
						ctx.Bool("endless"),
					)

					if err := crawler.Crawl(ctx.Context); err != nil {
						return fmt.Errorf("crawl submission: %w", err)
					}
					return nil
				},
			},
		},
	}
}

File: backend/src/cmd/crawl.rs
use std::time::Duration;

use crate::{
    atcoder::{AtCoderClient, AtCoderProblemsClient},
    crawl::{contest, difficulty, problem, user},
};
use clap::Subcommand;
use sqlx::postgres::PgPoolOptions;

use super::CommonArgs;

#[derive(Subcommand)]
pub(crate) enum CrawlCommand {
    Problem {
        #[arg(long, default_value_t = 1, help = "crawl duration in sec.")]
        duration: u64,
        #[arg(short, long, help = "if true, crawl all problems.")]
        all: bool,
    },
    User {
        #[arg(long, default_value_t = 1)]
        duration: u64,
    },
    Submission {
        #[arg(long, default_value_t = 1)]
        duration: u64,
        #[arg(long, default_value_t = 0)]
        retry: i64,
        #[arg(long)]
        target: String,
        #[arg(long, env, hide_env_values = true)]
        atcoder_username: String,
        #[arg(long, env, hide_env_values = true)]
        atcoder_password: String,
    },
}

impl CrawlCommand {
    pub async fn exec(&self, args: &CommonArgs) -> anyhow::Result<()> {
        let problems_client = AtCoderProblemsClient::new()?;
        let atcoder_client = AtCoderClient::new()?;
        let pool = PgPoolOptions::new()
            .max_connections(8)
            .connect(&args.database_url)
            .await?;

        match self {
            Self::Problem { duration, all } => {
                contest::crawl_contests(&problems_client, &pool).await?;
                difficulty::crawl_difficulties(&problems_client, &pool).await?;
                problem::crawl_problems(
                    &atcoder_client,
                    &problems_client,
                    &pool,
                    *all,
                    Duration::from_secs(*duration),
                )
                .await?;

                Ok(())
            }
            Self::User { duration } => {
                user::crawl_users(&atcoder_client, &pool, Duration::from_secs(*duration)).await
            }
            Self::Submission {
                duration,
                retry,
                target,
                atcoder_username,
                atcoder_password,
            } => {
                todo!();
            }
        }
    }
}
